### PR TITLE
Constants and Paths

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

- set up a file with text constants
- switched over to using the Path module for quite a few of the file/path-related functions

# Jira Ticket / Issue #

N/A

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (Python) I have run Ruff on changed Python code
- [ ] If appropriate, I have recompiled the app and added the updated `StaticNarrativeImpl.py`, `StaticNarrativeServer.py`, and `compile_report.json` to this PR.

### CODE REVIEW

File: lib/StaticNarrative/exceptions.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+"""Custom StaticNarrative Errors."""
```
Comment (ialarmedalien) [inline lib/StaticNarrative/exceptions.py:1]: Adding some extra documentation to this file.

File: lib/StaticNarrative/narrative_ref.py [hunk 2]
```diff
@@ -8,8 +8,12 @@
 
 
 class NarrativeRef:
-    def _less_than_zero(self: "NarrativeRef", number: str | int, name: str) -> int:
+    """Lightweight class representing a KBase reference."""
+
+    def _less_than_zero(self: "NarrativeRef", number: str | int | None, name: str) -> int:
         err_msg = f"The Narrative {name} must be an integer > 0, not {number}"
+        if number is None or not str(number).isdigit():
```
Comment (ialarmedalien) [inline lib/StaticNarrative/narrative_ref.py:15]: Casting to int meant that if you put in values like `4.5`, they would get converted to `4` and the ref would appear valid, so I made this a bit more strict.

File: lib/StaticNarrative/narrative_ref.py [hunk 3]
```diff
@@ -32,7 +36,18 @@ def __init__(self: "NarrativeRef", ref: dict[str, str | int]) -> None:
...
+        # initialise these attributes to prevent errors showing up all over the codebase
+        self.wsid: int
+        self.objid: int
+        self.ver: int
+
         part_name = {"wsid": "workspace ID", "objid": "object ID", "ver": "version"}
+        # ensure all parts are available
+        missing_parts = [part for part in part_name if part not in ref]
+        if missing_parts:
+            msg = "Missing keys required to create a NarrativeRef: " + ", ".join(missing_parts)
```
Comment (ialarmedalien) [inline lib/StaticNarrative/narrative_ref.py:48]: ensure that it isn't possible to do `NarrativeRef({"wsid": 12345})` and get a valid NarrativeRef out

File: test/mocks.py [hunk 4]
```diff
@@ -100,11 +100,11 @@ def mock_adapter(request: requests.Request) -> requests.Response:
                 tag = params[0]["tag"]
                 ids = params[0]["ids"]
                 result = [_get_fake_nms_info(tag, ids)]
-            response._content = bytes(json.dumps({"result": result, "version": "1.1"}), "UTF-8")
+            response._content = bytes(json.dumps({"result": result, "version": "1.1"}), "UTF-8")  # noqa: SLF001
```
Comment (ialarmedalien) [inline test/mocks.py:103]: prevent warnings when accessing `response._content`

File: test/mocks.py [hunk 5]
```diff
@@ -192,37 +192,13 @@ def set_up_ok_mocks(
...
-def mock_ws_narrative_fetch_forbidden(narrative_ref):
-    pass
-
-
-def mock_ws_info(ws_id):
-    pass
-
-
-def mock_ws_info_unauth(ws_id):
-    pass
```
Comment (ialarmedalien) [inline test/mocks.py:216]: none of these are used

File: test/run_tests.sh [hunk 6]
```diff
@@ -7,9 +7,6 @@ current_dir=$(dirname "$(readlink -f "$0")")
 export KB_DEPLOYMENT_CONFIG="$current_dir"/deploy.cfg
 export PYTHONPATH="$current_dir"/../lib:"$PYTHONPATH"
 
-# ensure there's a valid scratch dir and an assets dir
-mkdir -p "$current_dir"/../scratch/nginx
-
```
Comment (ialarmedalien) [inline test/run_tests.sh:12]: no longer need this (forgot to delete it in a previous PR)

File: test/test_config.py [hunk 7]
```diff
@@ -26,28 +20,24 @@ def mock_os_access() -> Generator[MagicMock, Any, None]:
...
     mock_os_access.side_effect = lambda x, y: True
-    mock_os_path_abspath.side_effect = lambda x: f"/absolute/{x}"
 
+    def starts_with_slash(self) -> bool:
+        """Fake function to replace is_absolute and is_dir."""
+        return str(self).startswith("/")
+
+    monkeypatch.setattr(Path, "is_absolute", starts_with_slash)
+    monkeypatch.setattr(Path, "is_dir", starts_with_slash)
+    monkeypatch.setattr(Path, "resolve", absolutify)
```
Comment (ialarmedalien) [inline test/test_config.py:40]: replace the old mocks with these monkey patched versions (arguably a bit simpler?)

File: test/test_static_narrative_server.py [hunk 8]
```diff
@@ -7,8 +7,6 @@
 
 from test.mocks import set_up_ok_mocks
 
-USER_ID = "some_user"
```
Comment (ialarmedalien) [inline test/test_static_narrative_server.py:10]: switching out `USER_ID` for `fake_user` from the conftest

File: test/test_narrative_util.py [hunk 9]
```diff
@@ -17,8 +17,6 @@
 
 from test.mocks import mock_ws_bad, set_up_ok_mocks
 
-USER_ID = "some_user"
-TOKEN = "some_token"  # noqa: S105
```
Comment (ialarmedalien) [inline test/test_narrative_util.py:21]: unused

File: test/test_config.py [hunk 10]
```diff
@@ -1,23 +1,17 @@
...
-        yield mock_isabs
-
-
-@pytest.fixture()
-def mock_os_path_isdir() -> Generator[MagicMock, Any, None]:
-    with patch("os.path.isdir") as mock_isdir:
-        yield mock_isdir
+def absolutify(self: Path) -> Path:
+    """Generate an absolute version of a path."""
+    return Path("/absolute") / self
```
Comment (briehl) [inline test/test_config.py:14]: That's pretty absolute!

**Review by briehl — APPROVED** [review]

**Pull request merged**